photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala

```diff
@@ -104,28 +104,39 @@ object RandomEffectCoordinate {
       randomEffectOptimizationProblem: RandomEffectOptimizationProblem[Function],
       randomEffectModel: RandomEffectModel): (RandomEffectModel, Option[RandomEffectOptimizationTracker]) = {
 
-    val updatedModels = randomEffectDataSet
+    val updatedModelsAndTrackers = randomEffectDataSet
       .activeData
       .join(randomEffectOptimizationProblem.optimizationProblems)
       .join(randomEffectModel.modelsRDD)
       .mapValues {
         case (((localDataSet, optimizationProblem), localModel)) =>
           val trainingLabeledPoints = localDataSet.dataPoints.map(_._2)
+          val updatedModel = optimizationProblem.run(trainingLabeledPoints, localModel)
+          val stateTrackers = optimizationProblem.getStatesTracker
 
-          optimizationProblem.run(trainingLabeledPoints, localModel)
+          (updatedModel, stateTrackers)
       }
+      .setName(s"Updated models and state trackers for random effect ${randomEffectDataSet.randomEffectType}")
+      .persist(StorageLevel.VERY_FREQUENT_REUSE_RDD_STORAGE_LEVEL)
     val updatedRandomEffectModel = randomEffectModel
-      .update(updatedModels)
-      .setName(s"Updated random effect model")
-    val optimizationTracker = if (randomEffectOptimizationProblem.isTrackingState) {
-        val stateTrackers = randomEffectOptimizationProblem.optimizationProblems.map(_._2.getStatesTracker.get)
+      .update(updatedModelsAndTrackers.mapValues(_._1))
+      .setName(s"Updated models for random effect ${randomEffectDataSet.randomEffectType}")
+      .persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
+      .materialize()
+    val optimizationTracker: Option[RandomEffectOptimizationTracker] =
+      if (randomEffectOptimizationProblem.isTrackingState) {
+        val stateTrackers = updatedModelsAndTrackers.map(_._2._2.get)
         val randomEffectTracker = new RandomEffectOptimizationTracker(stateTrackers)
-          .setName(s"Random effect optimization tracker")
+          .setName(s"State trackers for random effect ${randomEffectDataSet.randomEffectType}")
           .persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
           .materialize()
 
         Some(randomEffectTracker)
-      } else None
+      } else {
+        None.asInstanceOf[Option[RandomEffectOptimizationTracker]]
+      }
+
+    updatedModelsAndTrackers.unpersist()
```

Comment:
I don't understand the question.

We're trying to produce (up to) two new objects, each of which owns an `RDD`. One owns an `RDD` of models, the other an `RDD` of states. These two `RDD`s share a common ancestor in their lineage, because the models and states are computed at the same time. We compute the ancestor `RDD` (which is most quickly computed now, while IT'S ancestors are also persisted), temporarily persist it, force computation of the two derived `RDD`s (while the ancestor `RDD` exists), and then unpersist the ancestor (so that we do not need to track it and unpersist it later).